src/Common/filesystemHelpers.cpp

```diff
@@ -234,34 +234,22 @@ bool createFile(const std::string & path)
     DB::throwFromErrnoWithPath("Cannot create file: " + path, path, DB::ErrorCodes::CANNOT_CREATE_FILE);
 }
 
+bool exists(const std::string & path)
+{
+    return faccessat(AT_FDCWD, path.c_str(), F_OK, AT_EACCESS) == 0;
+}
+
 bool canRead(const std::string & path)
 {
-    struct stat st;
-    if (stat(path.c_str(), &st) == 0)
-    {
-        if (st.st_uid == geteuid())
-            return (st.st_mode & S_IRUSR) != 0;
-        else if (st.st_gid == getegid())
-            return (st.st_mode & S_IRGRP) != 0;
-        else
-            return (st.st_mode & S_IROTH) != 0 || geteuid() == 0;
-    }
+    if (exists(path))
+        return faccessat(AT_FDCWD, path.c_str(), R_OK, AT_EACCESS) == 0;
```

Comment:
Instead of making 2 syscalls (one to check if file exists, another to check read/write access) it's possible to do one syscall with flags `F_OK | R_OK`, but I avoided that since then we can't throw the original exception: 
```cpp
DB::throwFromErrnoWithPath("Cannot check read access to file: " + path, path, DB::ErrorCodes::PATH_ACCESS_DENIED);
````
Is it important for us to keep that exception?

Reply:
Looks like in this case you can distinguish them by `errno`, `ENOENT` if file does not exists

Reply:
The change increases the number of heavy syscalls.
`exists(path)` is comparable to `stat`